### ISSUE TITLE
3.47: Tests fail to compile

### ISSUE THREAD

**Issue by yurivict** [stillwater-sc/universal#263]
clang-12 prints these errors:

```
In file included from /disk-samsung/freebsd-ports/math/universal/work/universal-3.47/tools/cmd/areal.cpp:6:
In file included from /disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/areal.hpp:37:
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:13:86: error: implicit instantiation of undefined template 'std::basic_string<char>'
        areal_arithmetic_exception(const std::string& err) : universal_arithmetic_exception(std::string("areal arithmetic exception: ") + err) {};
                                                                                            ^
/usr/include/c++/v1/iosfwd:239:32: note: template is declared here
    class _LIBCPP_TEMPLATE_VIS basic_string;
                               ^
In file included from /disk-samsung/freebsd-ports/math/universal/work/universal-3.47/tools/cmd/areal.cpp:6:
In file included from /disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/areal.hpp:37:
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:13:130: error: invalid operands to binary expression ('std::string' (aka 'basic_string<char>') and 'const std::string' (aka 'const basic_string<char>'))
        areal_arithmetic_exception(const std::string& err) : universal_arithmetic_exception(std::string("areal arithmetic exception: ") + err) {};
                                                                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:18:77: error: implicit instantiation of undefined template 'std::basic_string<char>'
        areal_quire_exception(const std::string& err) : areal_arithmetic_exception(std::string("areal quire exception: ") + err) {}
                                                                                   ^
/usr/include/c++/v1/iosfwd:239:32: note: template is declared here
    class _LIBCPP_TEMPLATE_VIS basic_string;
                               ^
In file included from /disk-samsung/freebsd-ports/math/universal/work/universal-3.47/tools/cmd/areal.cpp:6:
In file included from /disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/areal.hpp:37:
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:18:116: error: invalid operands to binary expression ('std::string' (aka 'basic_string<char>') and 'const std::string' (aka 'const basic_string<char>'))
        areal_quire_exception(const std::string& err) : areal_arithmetic_exception(std::string("areal quire exception: ") + err) {}
                                                                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:26:25: error: no matching constructor for initialization of 'sw::universal::areal_arithmetic_exception'
        areal_not_a_number() : areal_arithmetic_exception("not a number") {}
                               ^                          ~~~~~~~~~~~~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:12:8: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const char [13]' to 'const sw::universal::areal_arithmetic_exception' for 1st argument
struct areal_arithmetic_exception : public universal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:12:8: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const char [13]' to 'sw::universal::areal_arithmetic_exception' for 1st argument
struct areal_arithmetic_exception : public universal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:13:2: note: candidate constructor not viable: no known conversion from 'const char [13]' to 'const std::string' (aka 'const basic_string<char>') for 1st argument
        areal_arithmetic_exception(const std::string& err) : universal_arithmetic_exception(std::string("areal arithmetic exception: ") + err) {};
        ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:31:27: error: no matching constructor for initialization of 'sw::universal::areal_arithmetic_exception'
        areal_divide_by_zero() : areal_arithmetic_exception("divide by zero") {}
                                 ^                          ~~~~~~~~~~~~~~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:12:8: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const char [15]' to 'const sw::universal::areal_arithmetic_exception' for 1st argument
struct areal_arithmetic_exception : public universal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:12:8: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const char [15]' to 'sw::universal::areal_arithmetic_exception' for 1st argument
struct areal_arithmetic_exception : public universal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:13:2: note: candidate constructor not viable: no known conversion from 'const char [15]' to 'const std::string' (aka 'const basic_string<char>') for 1st argument
        areal_arithmetic_exception(const std::string& err) : universal_arithmetic_exception(std::string("areal arithmetic exception: ") + err) {};
        ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:36:26: error: no matching constructor for initialization of 'sw::universal::areal_arithmetic_exception'
        areal_divide_by_nan() : areal_arithmetic_exception("divide by nan") {}
                                ^                          ~~~~~~~~~~~~~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:12:8: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const char [14]' to 'const sw::universal::areal_arithmetic_exception' for 1st argument
struct areal_arithmetic_exception : public universal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:12:8: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const char [14]' to 'sw::universal::areal_arithmetic_exception' for 1st argument
struct areal_arithmetic_exception : public universal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:13:2: note: candidate constructor not viable: no known conversion from 'const char [14]' to 'const std::string' (aka 'const basic_string<char>') for 1st argument
        areal_arithmetic_exception(const std::string& err) : universal_arithmetic_exception(std::string("areal arithmetic exception: ") + err) {};
        ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:41:27: error: no matching constructor for initialization of 'sw::universal::areal_arithmetic_exception'
        areal_operand_is_nan() : areal_arithmetic_exception("operand is nan") {}
                                 ^                          ~~~~~~~~~~~~~~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:12:8: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const char [15]' to 'const sw::universal::areal_arithmetic_exception' for 1st argument
struct areal_arithmetic_exception : public universal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:12:8: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const char [15]' to 'sw::universal::areal_arithmetic_exception' for 1st argument
struct areal_arithmetic_exception : public universal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:13:2: note: candidate constructor not viable: no known conversion from 'const char [15]' to 'const std::string' (aka 'const basic_string<char>') for 1st argument
        areal_arithmetic_exception(const std::string& err) : universal_arithmetic_exception(std::string("areal arithmetic exception: ") + err) {};
        ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:46:30: error: no matching constructor for initialization of 'sw::universal::areal_arithmetic_exception'
        areal_negative_sqrt_arg() : areal_arithmetic_exception("negative sqrt argument") {}
                                    ^                          ~~~~~~~~~~~~~~~~~~~~~~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:12:8: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const char [23]' to 'const sw::universal::areal_arithmetic_exception' for 1st argument
struct areal_arithmetic_exception : public universal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:12:8: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const char [23]' to 'sw::universal::areal_arithmetic_exception' for 1st argument
struct areal_arithmetic_exception : public universal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:13:2: note: candidate constructor not viable: no known conversion from 'const char [23]' to 'const std::string' (aka 'const basic_string<char>') for 1st argument
        areal_arithmetic_exception(const std::string& err) : universal_arithmetic_exception(std::string("areal arithmetic exception: ") + err) {};
        ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:51:33: error: no matching constructor for initialization of 'sw::universal::areal_quire_exception'
        areal_quire_operand_is_nan() : areal_quire_exception("quire operand is nan") {}
                                       ^                     ~~~~~~~~~~~~~~~~~~~~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:17:8: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const char [21]' to 'const sw::universal::areal_quire_exception' for 1st argument
struct areal_quire_exception : public areal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:17:8: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const char [21]' to 'sw::universal::areal_quire_exception' for 1st argument
struct areal_quire_exception : public areal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:18:2: note: candidate constructor not viable: no known conversion from 'const char [21]' to 'const std::string' (aka 'const basic_string<char>') for 1st argument
        areal_quire_exception(const std::string& err) : areal_arithmetic_exception(std::string("areal quire exception: ") + err) {}
        ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:56:36: error: no matching constructor for initialization of 'sw::universal::areal_quire_exception'
        areal_quire_negative_sqrt_arg() : areal_quire_exception("quire negative sqrt argument") {}
                                          ^                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:17:8: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const char [29]' to 'const sw::universal::areal_quire_exception' for 1st argument
struct areal_quire_exception : public areal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:17:8: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const char [29]' to 'sw::universal::areal_quire_exception' for 1st argument
struct areal_quire_exception : public areal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:18:2: note: candidate constructor not viable: no known conversion from 'const char [29]' to 'const std::string' (aka 'const basic_string<char>') for 1st argument
        areal_quire_exception(const std::string& err) : areal_arithmetic_exception(std::string("areal quire exception: ") + err) {}
        ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:63:82: error: implicit instantiation of undefined template 'std::basic_string<char>'
        areal_internal_exception(const std::string& err) : universal_internal_exception(std::string("areal internal exception: ") + err) {};
                                                                                        ^
/usr/include/c++/v1/iosfwd:239:32: note: template is declared here
    class _LIBCPP_TEMPLATE_VIS basic_string;
                               ^
In file included from /disk-samsung/freebsd-ports/math/universal/work/universal-3.47/tools/cmd/areal.cpp:6:
In file included from /disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/areal.hpp:37:
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:63:124: error: invalid operands to binary expression ('std::string' (aka 'basic_string<char>') and 'const std::string' (aka 'const basic_string<char>'))
        areal_internal_exception(const std::string& err) : universal_internal_exception(std::string("areal internal exception: ") + err) {};
                                                                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:67:28: error: no matching constructor for initialization of 'sw::universal::areal_internal_exception'
        areal_shift_too_large() : areal_internal_exception("shift value too large for given areal") {}
                                  ^                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:62:8: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const char [38]' to 'const sw::universal::areal_internal_exception' for 1st argument
struct areal_internal_exception : public universal_internal_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:62:8: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const char [38]' to 'sw::universal::areal_internal_exception' for 1st argument
struct areal_internal_exception : public universal_internal_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:63:2: note: candidate constructor not viable: no known conversion from 'const char [38]' to 'const std::string' (aka 'const basic_string<char>') for 1st argument
        areal_internal_exception(const std::string& err) : universal_internal_exception(std::string("areal internal exception: ") + err) {};
        ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:71:27: error: no matching constructor for initialization of 'sw::universal::areal_internal_exception'
        areal_hpos_too_large() : areal_internal_exception("position of hidden bit too large for given areal") {}
                                 ^                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:62:8: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const char [49]' to 'const sw::universal::areal_internal_exception' for 1st argument
struct areal_internal_exception : public universal_internal_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:62:8: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const char [49]' to 'sw::universal::areal_internal_exception' for 1st argument
struct areal_internal_exception : public universal_internal_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/areal/exceptions.hpp:63:2: note: candidate constructor not viable: no known conversion from 'const char [49]' to 'const std::string' (aka 'const basic_string<char>') for 1st argument
        areal_internal_exception(const std::string& err) : universal_internal_exception(std::string("areal internal exception: ") + err) {};
        ^
15 errors generated.
[  4% 18/216] /usr/bin/c++ -DGIT_BRANCH=main -DGIT_COMMIT_HASH=81327a9053c5 -I/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include -O2 -pipe -fno-omit-frame-pointer -fstack-protector-strong -fno-strict-aliasing -fno-omit-frame-pointer   -Wall -Wpedantic -Wno-narrowing -Wno-deprecated -O2 -pipe -fno-omit-frame-pointer -fstack-protector-strong -fno-strict-aliasing -fno-omit-frame-pointer   -Wall -Wpedantic -Wno-narrowing -Wno-deprecated    -D REGRESSION_LEVEL_OVERRIDE -D REGRESSION_LEVEL_1=1 -D REGRESSION_LEVEL_2=0 -D REGRESSION_LEVEL_3=0 -D REGRESSION_LEVEL_4=0 -std=gnu++20 -MD -MT tools/cmd/CMakeFiles/lns.dir/lns.cpp.o -MF tools/cmd/CMakeFiles/lns.dir/lns.cpp.o.d -o tools/cmd/CMakeFiles/lns.dir/lns.cpp.o -c /disk-samsung/freebsd-ports/math/universal/work/universal-3.47/tools/cmd/lns.cpp
FAILED: tools/cmd/CMakeFiles/lns.dir/lns.cpp.o 
/usr/bin/c++ -DGIT_BRANCH=main -DGIT_COMMIT_HASH=81327a9053c5 -I/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include -O2 -pipe -fno-omit-frame-pointer -fstack-protector-strong -fno-strict-aliasing -fno-omit-frame-pointer   -Wall -Wpedantic -Wno-narrowing -Wno-deprecated -O2 -pipe -fno-omit-frame-pointer -fstack-protector-strong -fno-strict-aliasing -fno-omit-frame-pointer   -Wall -Wpedantic -Wno-narrowing -Wno-deprecated    -D REGRESSION_LEVEL_OVERRIDE -D REGRESSION_LEVEL_1=1 -D REGRESSION_LEVEL_2=0 -D REGRESSION_LEVEL_3=0 -D REGRESSION_LEVEL_4=0 -std=gnu++20 -MD -MT tools/cmd/CMakeFiles/lns.dir/lns.cpp.o -MF tools/cmd/CMakeFiles/lns.dir/lns.cpp.o.d -o tools/cmd/CMakeFiles/lns.dir/lns.cpp.o -c /disk-samsung/freebsd-ports/math/universal/work/universal-3.47/tools/cmd/lns.cpp
In file included from /disk-samsung/freebsd-ports/math/universal/work/universal-3.47/tools/cmd/lns.cpp:6:
In file included from /disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/lns.hpp:37:
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/exceptions.hpp:12:84: error: implicit instantiation of undefined template 'std::basic_string<char>'
        lns_arithmetic_exception(const std::string& err) : universal_arithmetic_exception(std::string("lns arithmetic exception: ") + err) {}
                                                                                          ^
/usr/include/c++/v1/iosfwd:239:32: note: template is declared here
    class _LIBCPP_TEMPLATE_VIS basic_string;
                               ^
In file included from /disk-samsung/freebsd-ports/math/universal/work/universal-3.47/tools/cmd/lns.cpp:6:
In file included from /disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/lns.hpp:37:
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/exceptions.hpp:12:126: error: invalid operands to binary expression ('std::string' (aka 'basic_string<char>') and 'const std::string' (aka 'const basic_string<char>'))
        lns_arithmetic_exception(const std::string& err) : universal_arithmetic_exception(std::string("lns arithmetic exception: ") + err) {}
                                                                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/exceptions.hpp:17:25: error: no matching constructor for initialization of 'sw::universal::lns_arithmetic_exception'
        lns_divide_by_zero() : lns_arithmetic_exception("division by zero") {}
                               ^                        ~~~~~~~~~~~~~~~~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/exceptions.hpp:11:8: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const char [17]' to 'const sw::universal::lns_arithmetic_exception' for 1st argument
struct lns_arithmetic_exception : public universal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/exceptions.hpp:11:8: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const char [17]' to 'sw::universal::lns_arithmetic_exception' for 1st argument
struct lns_arithmetic_exception : public universal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/exceptions.hpp:12:2: note: candidate constructor not viable: no known conversion from 'const char [17]' to 'const std::string' (aka 'const basic_string<char>') for 1st argument
        lns_arithmetic_exception(const std::string& err) : universal_arithmetic_exception(std::string("lns arithmetic exception: ") + err) {}
        ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/exceptions.hpp:22:28: error: no matching constructor for initialization of 'sw::universal::lns_arithmetic_exception'
        lns_negative_sqrt_arg() : lns_arithmetic_exception("negative sqrt argument") {}
                                  ^                        ~~~~~~~~~~~~~~~~~~~~~~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/exceptions.hpp:11:8: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const char [23]' to 'const sw::universal::lns_arithmetic_exception' for 1st argument
struct lns_arithmetic_exception : public universal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/exceptions.hpp:11:8: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const char [23]' to 'sw::universal::lns_arithmetic_exception' for 1st argument
struct lns_arithmetic_exception : public universal_arithmetic_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/exceptions.hpp:12:2: note: candidate constructor not viable: no known conversion from 'const char [23]' to 'const std::string' (aka 'const basic_string<char>') for 1st argument
        lns_arithmetic_exception(const std::string& err) : universal_arithmetic_exception(std::string("lns arithmetic exception: ") + err) {}
        ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/exceptions.hpp:31:34: error: implicit instantiation of undefined template 'std::basic_string<char>'
                : universal_internal_exception(std::string("lns internal error: ") + error) {};
                                               ^
/usr/include/c++/v1/iosfwd:239:32: note: template is declared here
    class _LIBCPP_TEMPLATE_VIS basic_string;
                               ^
In file included from /disk-samsung/freebsd-ports/math/universal/work/universal-3.47/tools/cmd/lns.cpp:6:
In file included from /disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/lns.hpp:37:
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/exceptions.hpp:31:70: error: invalid operands to binary expression ('std::string' (aka 'basic_string<char>') and 'const std::string' (aka 'const basic_string<char>'))
                : universal_internal_exception(std::string("lns internal error: ") + error) {};
                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/exceptions.hpp:35:30: error: no matching constructor for initialization of 'sw::universal::lns_internal_exception'
        lns_index_out_of_bounds() : lns_internal_exception("index out of bounds") {}
                                    ^                      ~~~~~~~~~~~~~~~~~~~~~
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/exceptions.hpp:29:8: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const char [20]' to 'const sw::universal::lns_internal_exception' for 1st argument
struct lns_internal_exception : public universal_internal_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/exceptions.hpp:29:8: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const char [20]' to 'sw::universal::lns_internal_exception' for 1st argument
struct lns_internal_exception : public universal_internal_exception {
       ^
/disk-samsung/freebsd-ports/math/universal/work/universal-3.47/./include/universal/number/lns/exceptions.hpp:30:2: note: candidate constructor not viable: no known conversion from 'const char [20]' to 'const std::string' (aka 'const basic_string<char>') for 1st argument
        lns_internal_exception(const std::string& error)
        ^
7 errors generated.
```

OS: FreeBSD 13

**Comment by Ravenwater** [comment]
Acknowledged. I'll clean that up in v3.48.

**Comment by Ravenwater** [comment]
#include <string> was missing in the exception include hierarchy.

fixed in v3.48, which is in CI right now. Will release when clean.
